isovar/assembly.py

```diff
@@ -16,56 +16,48 @@
 
 from collections import defaultdict
 import logging
-# I'd rather just use list.copy but Python 2.7 doesn't support it
-from copy import copy
 
-from .default_parameters import MIN_VARIANT_CDNA_SEQUENCE_ASSEMBLY_OVERLAP_SIZE
+from .default_parameters import MIN_VARIANT_SEQUENCE_ASSEMBLY_OVERLAP_SIZE
 
 logger = logging.getLogger(__name__)
 
-def sort_by_decreasing_prefix_length(seq):
-    """
-    Key function for sorting from longest to shortest prefix length.
-
-    Parameters
-    ----------
-    seq : VariantSequence
-    """
-    return -len(seq.prefix)
-
-def sort_by_decreasing_suffix_length(seq):
-    """
-    Key function for sorting from longest to shortest suffix length.
-
-    Parameters
-    ----------
-    seq : VariantSequence
-    """
-    return -len(seq.suffix)
-
-def sort_by_increasing_total_length(seq):
-    """
-    Key function for sorting from shortest to longest total length.
-
-    Parameters
-    ----------
-    seq : VariantSequence
-    """
-    return len(seq.sequence)
-
-def sort_by_decreasing_total_length(seq):
-    """
-    Key function for sorting from longest to shortest total length.
-
-    Parameters
-    ----------
-    seq : VariantSequence
+def greedy_merge_helper(
+        variant_sequences,
+        min_overlap_size=MIN_VARIANT_SEQUENCE_ASSEMBLY_OVERLAP_SIZE):
     """
-    return -len(seq.sequence)
+    Returns a list of merged VariantSequence objects, and True if any
+    were successfully merged.
+    """ 
+    merged_variant_sequences = {}
+    merged_any = False
+
+    # here we'll keep track of sequences that haven't been merged yet, and add them in at the end
+    unmerged_variant_sequences = set(variant_sequences)
+    for i in range(len(variant_sequences)):
+        sequence1 = variant_sequences[i]
+        for j in range(i+1, len(variant_sequences)):
+            sequence2 = variant_sequences[j]
+            try:
+                s = sequence1.combine(sequence2)
+                if s.sequence in merged_variant_sequences:
+                    existing_s = merged_variant_sequences[s.sequence]
+                    # we may have already created the same sequence from another set of reads, in
+                    # which case we need to merge the reads
+                    if s.reads != existing_s.reads:
+                        new_s = existing_s.combine(s)
+                        s = new_s
+                merged_variant_sequences[s.sequence] = s
+                unmerged_variant_sequences.discard(sequence1)
+                unmerged_variant_sequences.discard(sequence2)
+                merged_any = True
+            except ValueError:
```

Comment:
...since catching ValueError will suppress other exceptions which are unrelated to `combine`.

Alternatively I could make a custom exception class like `CombineSequencesFailed`.